src/libsyntax_ext/concat.rs

```diff
@@ -53,7 +54,13 @@ pub fn expand_syntax_ext(cx: &mut base::ExtCtxt,
                 }
             }
             _ => {
-                cx.span_err(e.span, "expected a literal");
+                let mut err = cx.struct_span_err(e.span, "expected a literal");
+                err.span_suggestion(
+                    e.span,
+                    "consider changing this to",
+                    format!("\"{{}}\", {}", pprust::expr_to_string(&e))
```

Comment:
Could you use span to snippet here and only use expression to string only if that call fails? We want to preserve the code as is, like for example whitespace.